File: InterfaceBuilding/SwiftInterfaceBulding/SwiftInterfaceBulding/Tester.swift
//
//  tester.swift
//  SwiftInterfaceBulding
//
//  Created by Krzysztof Drab on 29/12/2020.
//  Copyright © 2020 Krzysztof Drab. All rights reserved.
//

import Foundation

struct TestResult {
    var n: Int
    var timeToModify: Double
    var fps: Double
}

struct RowData: Identifiable {
    var id: Int
//    var id = UUID() It is slower because comparison is more expensive
    var value: Double
    var visibility = true
}

class Tester {
    var results: [TestResult] = [];
    var tests: [Int] = [];
    var timeToModify: Double = 0.0;
    var testId: Int = -1;

    func generateRows() -> [RowData] {
        preconditionFailure("This method must be overridden")
    }
    func updateRows() -> [RowData] {
        preconditionFailure("This method must be overridden")
    }
    func testEnded(fps: Double) {
        preconditionFailure("This method must be overridden")
    }

    func generateDefaultRows(nbOfRows: Int) -> [RowData] {
        let arr = Array<Double>(repeating: 0.0, count: nbOfRows)
        return arr.map { _ in
            RowData(id: Int.random(in: 0...10000), value: Double.random(in: 0...1))
        };
    }

    func nextTest() -> [RowData] {
      self.testId += 1;
      self.timeToModify = 0.0;
      return self.generateRows();
    }

    func isCompleted() -> Bool {
      return self.testId >= self.tests.count - 1;
    }
}

File: InterfaceBuilding/flutter_interface_building/lib/tester.dart
import 'package:flutter_interface_building/row_data.dart';

class TestResult {
  int n;
  double fps;
  double timeToModify;

  TestResult([this.n, this.fps, this.timeToModify]);
}

abstract class Tester {
  List<TestResult> results = [];
  List<int> tests = [];
  double timeToModify;
  int testId = -1;

  List<RowData> generateRows();
  List<RowData> updateRows();
  void testEnded(double fps);

  List<RowData> generateDefaultRows(nbOfRows) {
    return new List(nbOfRows).map((e) => RowData()).toList();
  }

  List<RowData> nextTest() {
    this.testId += 1;
    this.timeToModify = 0.0;
    return this.generateRows();
  }

  bool isCompleted() {
    return this.testId >= this.tests.length - 1;
  }
}